### ISSUE TITLE
Bug: secondary redbean SQLite WAL connections lock after first write

### ISSUE THREAD

**Issue by goldenstein64** [jart/cosmopolitan#1228]
### Contact Details

goldnstein64@gmail.com

### What happened?

If two processes are connected to the same database with WAL enabled, a write operation on one connection will block all other connections until the writer closes.

Here is a reproduction of the issue along with its output, executed with `redbean -F FILE -i`

```lua
unix.rmrf('some.db')

do
  local ddl_db = assert(lsqlite3.open('some.db'))
  ddl_db:exec('PRAGMA journal_mode=WAL')
  ddl_db:close()
end

local pid = assert(unix.fork()) -- run this in 2 processes
local pname = pid == 0 and 'child:' or 'parent:'
local db = assert(lsqlite3.open('some.db'))

-- make the parent sleep for one second
if pid ~= 0 then
  unix.nanosleep(1)
end

print(pname, db:exec('PRAGMA optimize')) -- execute a write operation (I think, still blocks)
print(pname, db:error_message())

-- if executed successfully, sleep for 15 seconds
if db:error_code() == lsqlite3.OK then
  unix.nanosleep(15)
end
db:close()
unix.exit()
```

```
child:  0
child:  not an error
parent: 15
parent: locking protocol
```

The `parent:` messages appear after ~13 seconds. Making the write explicitly finish with `BEGIN TRANSACTION` / `END TRANSACTION` or `COMMIT` doesn't alter the program's behavior.

The expected output for this program would likely be:

```
child:  0
child:  not an error
parent: 0
parent: not an error
```

### Version

redbean 2.2

### What operating system are you seeing the problem on?

Windows

### Relevant log output

```shell
> ./redbean.com -F test.lua -i --strace
SYS  29048            714'348 bell system five system call support 430 magnums loaded on the new technology
SYS  29048            808'858 __morph_begin()
SYS  29048          1'433'336 __morph_end()

> ./redbean.com -F test.lua -i --ftrace
error: --ftrace failed to open symbol table
child:  0
child:  not an error
parent: 15
parent: locking protocol
```

**Comment by pkulchenko** [comment]
Thank you for the test case to reproduce the issue. I ran into the same issue some time ago and even posted my example to the SQLite forum, but given that cosmo implementation on Windows is different from the "original" sqlite one (as it's using the same code that runs on unix), there wasn't much response.

I can provide an additional case that may help with troubleshooting. Try commenting out the removal of the file on line one and then the "close" call will make all the difference between successful and failing cases:

```
--unix.rmrf('some.db')

do
  local ddl_db = assert(lsqlite3.open('some.db'))
  ddl_db:exec('PRAGMA journal_mode=WAL')
--  ddl_db:close() -- succeeds without close() and fails with close() call
end

local pid = assert(unix.fork()) -- run this in 2 processes
local pname = pid == 0 and 'child:' or 'parent:'
local db = assert(lsqlite3.open('some.db'))

-- make the parent sleep for one second
if pid ~= 0 then
  unix.nanosleep(1)
end

print(pname, db:exec('PRAGMA optimize')) -- execute a write operation (I think, still blocks)
print(pname, db:error_message())

-- if executed successfully, sleep for 15 seconds
if db:error_code() == lsqlite3.OK then
  unix.nanosleep(15)
end
db:close()

print(pname, "done")
if pid ~= 0 then unix.wait(pid) end -- added to reap zombies

unix.exit()
```

Somehow calling `close()` affects the structures that the process has before forking, which "breaks" the locking mechanism. I tried to troubleshoot it some time ago and even compare with the same code that succeeds on Linux, but couldn't find the culprit.

I'm open to ideas/suggestions, as it affects one of my projects.

**Comment by ncruces** [comment]
You're getting an [`SQLITE_PROTOCOL`](https://www.sqlite.org/rescode.html#protocol) error, which points to SQLite being blocked on a WAL lock that SQLite never expects to be held for very long (so the busy handler is not involved).

See [this](https://github.com/sqlite/sqlite/blob/51b3b402c49c2535749c7e790fb52a017e3f5674/src/wal.c#L3002-L3026) which is called in a loop [here](https://github.com/sqlite/sqlite/blob/51b3b402c49c2535749c7e790fb52a017e3f5674/src/wal.c#L3374-L3376). I'd look for places where `WAL_RETRY` is returned.

If this is happening when opening a new connection, I suspect the DMS lock might be involved. The Windows and Unix implementations are very different. Trying to do Windows by translating Unix syscalls is likely to introduce races here.

**Comment by goldenstein64** [comment]
I think the bug fixed itself in the latest version? :+1:?

**Comment by goldenstein64** [comment]
Nevermind, I just realized that `ddl_db:close()` was commented out in my version of the repro file... embarrassing

**Comment by pkulchenko** [comment]
> See [this](https://github.com/sqlite/sqlite/blob/51b3b402c49c2535749c7e790fb52a017e3f5674/src/wal.c#L3002-L3026) which is called in a loop [here](https://github.com/sqlite/sqlite/blob/51b3b402c49c2535749c7e790fb52a017e3f5674/src/wal.c#L3374-L3376). I'd look for places where WAL_RETRY is returned.
> If this is happening when opening a new connection, I suspect the DMS lock might be involved. The Windows and Unix implementations are very different. Trying to do Windows by translating Unix syscalls is likely to introduce races here.

@ncruces, thank you for the details. Do you have any further suggestions on what to check or how to troubleshoot this? It does work for me without closing the DB, so there is something that happens as part of this process that affects the subsequent processing, but I so far has not been able to find what that may be.

**Comment by ncruces** [comment]
No. I will note that what you're trying to do here is fraught with issues.

I hope you've read [sqlite.org/howtocorrupt.html](https://sqlite.org/howtocorrupt.html), especially the entire section 2; also [this](https://github.com/sqlite/sqlite/blob/b74eb0/src/os_unix.c#L1073-L1161).
